base/multidimensional.jl

```diff
@@ -823,14 +823,13 @@ julia> diff(vec(a))
 ```
 """
 function diff(a::AbstractArray{T,N}; dims::Integer) where {T,N}
-    require_one_based_indexing(a)
     1 <= dims <= N || throw(ArgumentError("dimension $dims out of range (1:$N)"))
 
     r = axes(a)
-    r0 = ntuple(i -> i == dims ? UnitRange(1, last(r[i]) - 1) : UnitRange(r[i]), N)
-    r1 = ntuple(i -> i == dims ? UnitRange(2, last(r[i])) : UnitRange(r[i]), N)
+    r0 = ntuple(i -> i == dims ? UnitRange(first(r[i]), last(r[i]) - 1) : UnitRange(r[i]), N)
+    r1 = ntuple(i -> i == dims ? UnitRange(first(r[i]) + 1, last(r[i])) : UnitRange(r[i]), N)
 
-    return view(a, r1...) .- view(a, r0...)
+    return reshape(view(a, r1...) .- view(a, r0...), r0)
```

Comment:
This `reshape` is causing test failures, e.g., `MethodError: no method matching reshape(::Array{Float64,2}, ::Tuple{UnitRange{Int64},UnitRange{Int64}})` (that method is defined only if you've loaded `OffsetArrays`).

In principle, you don't need `reshape` if you pay attention to the indices-of-the-indices. That is, `r0` and `r1` themselves have `Base.OneTo` axes, which is why the difference `view1 .- view2` has `Base.OneTo` axes and thus needs a `reshape`. But instead, if you make sure that they preserve their own axes, then the `view1 .- view2` will use those axes.

For example:
```julia
trimlast(ax::OneTo) = OneTo(ax.stop-1)
trimlast(ax::UnitRange) = UnitRange(ax.start, ax.stop-1)
trimlast(ax::IdentityUnitRange) = IdentityUnitRange(trimlast(ax.indices))

r0 = ntuple(i -> i == dims ? trimlast(r[i]) : r[i], N)
r1 = ntuple(i -> i == dims ? trimlast(r[i]) .+ 1 : r[i], N)

view(a, r1...) .- view(a, r0...)
```
seems like it should work, except that OffsetArrays will need to define that broadcasting operation for `IdentityUnitRange`.